Compiler error in 0.6.0.beta-3

To reproduce the issue I'm having, create a new crate with the following dependencies (possibly not all required)

```
[dependencies]
actix-web = "4.0.0-beta.8"
actix-cors = "0.6.0-beta.3"
actix-web-prom = "0.6.0-beta.3"
```
It gives the following compiler error:
```
error[E0432]: unresolved imports `actix_web::body::AnyBody`, `actix_web::dev::BodySize`, `actix_web::dev::MessageBody`, `actix_web::http::HeaderValue`
   --> /home/erlend/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-web-prom-0.6.0-beta.3/src/lib.rs:218:5
    |
218 |     body::AnyBody,
    |     ^^^^^^^^^^^^^ no `AnyBody` in `body`
219 |     dev::{BodySize, MessageBody, Service, ServiceRequest, ServiceResponse, Transform},
    |           ^^^^^^^^  ^^^^^^^^^^^ no `MessageBody` in `dev`
    |           |
    |           no `BodySize` in `dev`
220 |     http::{header::CONTENT_TYPE, HeaderValue, Method, StatusCode},
    |                                  ^^^^^^^^^^^ no `HeaderValue` in `http`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `actix-web-prom` due to previous error
```

rustc:
```
rustc 1.58.0-nightly (dd549dcab 2021-11-25)
```

Alright I got the dependency puzzle working.
Now I have one problem with prometheus:
```
error[E0271]: type mismatch resolving `<<impl ServiceFactory<ServiceRequest> as ServiceFactory<ServiceRequest>>::Service as actix_web::dev::Service<ServiceRequest>>::Response == ServiceResponse`
   --> src/main.rs:194:19
    |
194 |             .wrap(prometheus.clone())
    |              ---- ^^^^^^^^^^^^^^^^^^ expected enum `EitherBody`, found struct `BoxBody`
    |              |
    |              required by a bound introduced by this call
    |
    = note: expected struct `ServiceResponse<EitherBody<BoxBody>>`
               found struct `ServiceResponse<BoxBody>`
    = note: required because of the requirements on the impl of `Transform<<impl ServiceFactory<ServiceRequest> as ServiceFactory<ServiceRequest>>::Service, ServiceRequest>` for `PrometheusMetrics`
```

(where `prometheus: PrometheusMetrics`)

Relevant dependencies:
```toml
actix-web = "=4.0.0-beta.14"
actix-http = "=3.0.0-beta.15" # Only in order to pin the dep - TODO remove when actix-web 4 is released
actix-web-prom = { git = "https://github.com/nlopes/actix-web-prom.git", branch = "nlopes/bump-to-actix-web-14" }
prometheus = "0.13.0"
```

The examples run fine, but in my application I get 

```
error[E0271]: type mismatch resolving `<<impl ServiceFactory<ServiceRequest> as ServiceFactory<ServiceRequest>>::Service as actix_web::dev::Service<ServiceRequest>>::Response == ServiceResponse`
   --> src/main.rs:194:19
    |
194 |             .wrap(prometheus.clone())
    |              ---- ^^^^^^^^^^^^^^^^^^ expected enum `EitherBody`, found struct `BoxBody`
    |              |
    |              required by a bound introduced by this call
    |
    = note: expected struct `ServiceResponse<EitherBody<BoxBody>>`
               found struct `ServiceResponse<BoxBody>`
    = note: required because of the requirements on the impl of `Transform<<impl ServiceFactory<ServiceRequest> as ServiceFactory<ServiceRequest>>::Service, ServiceRequest>` for `PrometheusMetrics`
```
I tried my best to lock the right versions of all relevant deps:
```toml
actix-web = "=4.0.0-beta.16"
actix-http = "=3.0.0-beta.17"
actix-web-prom = { git = "https://github.com/nlopes/actix-web-prom.git", branch = "nlopes/bump-to-actix-web-14" }
prometheus = "0.13.0"
```

Any idea how I can fix this?

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -16,7 +16,7 @@ exclude = [".gitignore", ".travis.yml"]
 travis-ci = { repository = "nlopes/actix-web-prom", branch = "master" }
 
 [dependencies]
-actix-web = { version = "=4.0.0-beta.21", default-features = false }
-futures = "^0.3"
+actix-web = { version = "4.0.0-beta.21", default-features = false }
+futures-core = "0.3"
 pin-project = "^1.0"
 prometheus = { version = "^0.13", default-features = false }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -3,7 +3,7 @@ name = "actix-web-prom"
 version = "0.6.0-beta.5"
 authors = ["Norberto Lopes <nlopes.ml@gmail.com>"]
 edition = "2021"
-description = "Actix-web middleware to expose Prometheus metrics"
+description = "Actix Web middleware to expose Prometheus metrics"
 documentation = "https://docs.rs/actix-web-prom"
 repository = "https://github.com/nlopes/actix-web-prom.git"
 keywords = ["actix", "actix-web", "middleware", "prometheus", "web"]
@@ -18,5 +18,5 @@ travis-ci = { repository = "nlopes/actix-web-prom", branch = "master" }
 [dependencies]
 actix-web = { version = "4.0.0-beta.21", default-features = false }
 futures-core = "0.3"
-pin-project = "^1.0"
-prometheus = { version = "^0.13", default-features = false }
+pin-project-lite = "0.2"
+prometheus = { version = "0.13", default-features = false }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -210,13 +210,16 @@ fn main() -> std::io::Result<()> {
 #![deny(missing_docs)]
 
 use std::collections::HashMap;
+use std::future::{ready, Future, Ready};
 use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 use std::time::Instant;
 
+use actix_web::body::EitherBody;
 use actix_web::{
     body::{BodySize, BoxBody, MessageBody},
-    dev::{Service, ServiceRequest, ServiceResponse, Transform},
+    dev::{self, Service, ServiceRequest, ServiceResponse, Transform},
     http::{
         header::{HeaderValue, CONTENT_TYPE},
         Method, StatusCode,
@@ -224,12 +227,7 @@ use actix_web::{
     web::Bytes,
     Error,
 };
-use futures::{
-    future::{ok, Ready},
-    ready,
-    task::{Context, Poll},
-    Future,
-};
+use futures_core::ready;
 use prometheus::{
     Encoder, HistogramOpts, HistogramVec, IntCounterVec, Opts, Registry, TextEncoder,
 };
@@ -382,21 +380,21 @@ impl PrometheusMetrics {
     }
 }
 
-impl<S> Transform<S, ServiceRequest> for PrometheusMetrics
+impl<S, B> Transform<S, ServiceRequest> for PrometheusMetrics
 where
-    S: Service<ServiceRequest, Response = ServiceResponse, Error = Error>,
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Response = ServiceResponse;
+    type Response = ServiceResponse<EitherBody<StreamLog<B>>>;
     type Error = Error;
     type InitError = ();
     type Transform = PrometheusMetricsMiddleware<S>;
     type Future = Ready<Result<Self::Transform, Self::InitError>>;
 
     fn new_transform(&self, service: S) -> Self::Future {
-        ok(PrometheusMetricsMiddleware {
+        ready(Ok(PrometheusMetricsMiddleware {
             service,
             inner: Arc::new(self.clone()),
-        })
+        }))
     }
 }
 
@@ -413,16 +411,16 @@ where
     _t: PhantomData<()>,
 }
 
-impl<S> Future for LoggerResponse<S>
+impl<S, B> Future for LoggerResponse<S>
 where
-    S: Service<ServiceRequest, Response = ServiceResponse, Error = Error>,
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Output = Result<ServiceResponse, Error>;
+    type Output = Result<ServiceResponse<EitherBody<StreamLog<B>>>, Error>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let this = self.project();
 
-        let res = match futures::ready!(this.fut.poll(cx)) {
+        let res = match ready!(this.fut.poll(cx)) {
             Ok(res) => res,
             Err(e) => return Poll::Ready(Err(e)),
         };
@@ -436,7 +434,7 @@ where
         let path = req.path().to_string();
         let inner = this.inner.clone();
 
-        Poll::Ready(Ok(res.map_body(move |mut head, mut body| {
+        Poll::Ready(Ok(res.map_body(move |mut head, body| {
             // We short circuit the response status and body to serve the endpoint
             // automagically. This way the user does not need to set the middleware *AND*
             // an endpoint to serve middleware results. The user is only required to set
@@ -447,17 +445,27 @@ where
                     CONTENT_TYPE,
                     HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
                 );
-                body = BoxBody::new(inner.metrics());
+
+                EitherBody::right(BoxBody::new(StreamLog {
+                    body: inner.metrics(),
+                    size: 0,
+                    clock: time,
+                    inner,
+                    status: head.status,
+                    path: pattern_or_path,
+                    method,
+                }))
+            } else {
+                EitherBody::left(StreamLog {
+                    body,
+                    size: 0,
+                    clock: time,
+                    inner,
+                    status: head.status,
+                    path: pattern_or_path,
+                    method,
+                })
             }
-            BoxBody::new(StreamLog {
-                body,
-                size: 0,
-                clock: time,
-                inner,
-                status: head.status,
-                path: pattern_or_path,
-                method,
-            })
         })))
     }
 }
@@ -469,17 +477,15 @@ pub struct PrometheusMetricsMiddleware<S> {
     inner: Arc<PrometheusMetrics>,
 }
 
-impl<S> Service<ServiceRequest> for PrometheusMetricsMiddleware<S>
+impl<S, B> Service<ServiceRequest> for PrometheusMetricsMiddleware<S>
 where
-    S: Service<ServiceRequest, Response = ServiceResponse, Error = Error>,
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Response = ServiceResponse;
+    type Response = ServiceResponse<EitherBody<StreamLog<B>>>;
     type Error = S::Error;
     type Future = LoggerResponse<S>;
 
-    fn poll_ready(&self, ct: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.service.poll_ready(ct)
-    }
+    dev::forward_ready!(service);
 
     fn call(&self, req: ServiceRequest) -> Self::Future {
         LoggerResponse {
@@ -496,9 +502,9 @@ use std::marker::PhantomData;
 
 #[doc(hidden)]
 #[pin_project(PinnedDrop)]
-pub struct StreamLog {
+pub struct StreamLog<B> {
     #[pin]
-    body: BoxBody,
+    body: B,
     size: usize,
     clock: Instant,
     inner: Arc<PrometheusMetrics>,
@@ -508,7 +514,7 @@ pub struct StreamLog {
 }
 
 #[pinned_drop]
-impl PinnedDrop for StreamLog {
+impl<B> PinnedDrop for StreamLog<B> {
     fn drop(self: Pin<&mut Self>) {
         // update the metrics for this request at the very end of responding
         self.inner
@@ -516,8 +522,8 @@ impl PinnedDrop for StreamLog {
     }
 }
 
-impl MessageBody for StreamLog {
-    type Error = Error;
+impl<B: MessageBody> MessageBody for StreamLog<B> {
+    type Error = B::Error;
 
     fn size(&self) -> BodySize {
         self.body.size()
@@ -533,7 +539,7 @@ impl MessageBody for StreamLog {
                 *this.size += chunk.len();
                 Poll::Ready(Some(Ok(chunk)))
             }
-            Some(Err(err)) => Poll::Ready(Some(Err(err.into()))),
+            Some(Err(err)) => Poll::Ready(Some(Err(err))),
             None => Poll::Ready(None),
         }
     }
@@ -543,7 +549,7 @@ impl MessageBody for StreamLog {
 mod tests {
     use super::*;
     use actix_web::test::{call_and_read_body, call_service, init_service, read_body, TestRequest};
-    use actix_web::{web, App, HttpResponse};
+    use actix_web::{web, App, HttpResponse, Resource, Scope};
 
     use prometheus::{Counter, Opts};
 
@@ -931,5 +937,26 @@ actix_web_prom_http_requests_total{endpoint=\"/health_check\",label1=\"value1\",
             )
             .unwrap()
         ));
+    }
+
+    #[test]
+    fn compat_with_non_boxed_middleware() {
+        let _app = App::new()
+            .wrap(PrometheusMetricsBuilder::new("").build().unwrap())
+            .wrap(actix_web::middleware::Logger::default())
+            .route("", web::to(|| async { "" }));
+
+        let _app = App::new()
+            .wrap(actix_web::middleware::Logger::default())
+            .wrap(PrometheusMetricsBuilder::new("").build().unwrap())
+            .route("", web::to(|| async { "" }));
+
+        let _scope = Scope::new("")
+            .wrap(PrometheusMetricsBuilder::new("").build().unwrap())
+            .route("", web::to(|| async { "" }));
+
+        let _resource = Resource::new("")
+            .wrap(PrometheusMetricsBuilder::new("").build().unwrap())
+            .route(web::to(|| async { "" }));
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -217,9 +217,8 @@ use std::sync::Arc;
 use std::task::{Context, Poll};
 use std::time::Instant;
 
-use actix_web::body::EitherBody;
 use actix_web::{
-    body::{BodySize, BoxBody, MessageBody},
+    body::{BodySize, EitherBody, MessageBody},
     dev::{self, Service, ServiceRequest, ServiceResponse, Transform},
     http::{
         header::{HeaderValue, CONTENT_TYPE},
@@ -386,7 +385,7 @@ impl<S, B> Transform<S, ServiceRequest> for PrometheusMetrics
 where
     S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Response = ServiceResponse<EitherBody<StreamLog<B>>>;
+    type Response = ServiceResponse<EitherBody<StreamLog<B>, StreamLog<String>>>;
     type Error = Error;
     type InitError = ();
     type Transform = PrometheusMetricsMiddleware<S>;
@@ -418,7 +417,7 @@ impl<S, B> Future for LoggerResponse<S>
 where
     S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Output = Result<ServiceResponse<EitherBody<StreamLog<B>>>, Error>;
+    type Output = Result<ServiceResponse<EitherBody<StreamLog<B>, StreamLog<String>>>, Error>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let this = self.project();
@@ -449,7 +448,7 @@ where
                     HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
                 );
 
-                EitherBody::right(BoxBody::new(StreamLog {
+                EitherBody::right(StreamLog {
                     body: inner.metrics(),
                     size: 0,
                     clock: time,
@@ -457,7 +456,7 @@ where
                     status: head.status,
                     path: pattern_or_path,
                     method,
-                }))
+                })
             } else {
                 EitherBody::left(StreamLog {
                     body,
@@ -484,7 +483,7 @@ impl<S, B> Service<ServiceRequest> for PrometheusMetricsMiddleware<S>
 where
     S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
 {
-    type Response = ServiceResponse<EitherBody<StreamLog<B>>>;
+    type Response = ServiceResponse<EitherBody<StreamLog<B>, StreamLog<String>>>;
     type Error = S::Error;
     type Future = LoggerResponse<S>;
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -211,6 +211,7 @@ fn main() -> std::io::Result<()> {
 
 use std::collections::HashMap;
 use std::future::{ready, Future, Ready};
+use std::marker::PhantomData;
 use std::pin::Pin;
 use std::sync::Arc;
 use std::task::{Context, Poll};
@@ -228,6 +229,7 @@ use actix_web::{
     Error,
 };
 use futures_core::ready;
+use pin_project_lite::pin_project;
 use prometheus::{
     Encoder, HistogramOpts, HistogramVec, IntCounterVec, Opts, Registry, TextEncoder,
 };
@@ -235,7 +237,7 @@ use prometheus::{
 #[derive(Debug)]
 /// Builder to create new PrometheusMetrics struct.HistogramVec
 ///
-/// It allows settting optional parameters like registry, buckets, etc.
+/// It allows setting optional parameters like registry, buckets, etc.
 pub struct PrometheusMetricsBuilder {
     namespace: String,
     endpoint: Option<String>,
@@ -398,17 +400,18 @@ where
     }
 }
 
-#[doc(hidden)]
-#[pin_project::pin_project]
-pub struct LoggerResponse<S>
-where
-    S: Service<ServiceRequest>,
-{
-    #[pin]
-    fut: S::Future,
-    time: Instant,
-    inner: Arc<PrometheusMetrics>,
-    _t: PhantomData<()>,
+pin_project! {
+    #[doc(hidden)]
+    pub struct LoggerResponse<S>
+        where
+        S: Service<ServiceRequest>,
+    {
+        #[pin]
+        fut: S::Future,
+        time: Instant,
+        inner: Arc<PrometheusMetrics>,
+        _t: PhantomData<()>,
+    }
 }
 
 impl<S, B> Future for LoggerResponse<S>
@@ -497,28 +500,26 @@ where
     }
 }
 
-use pin_project::{pin_project, pinned_drop};
-use std::marker::PhantomData;
+pin_project! {
+    #[doc(hidden)]
+    pub struct StreamLog<B> {
+        #[pin]
+        body: B,
+        size: usize,
+        clock: Instant,
+        inner: Arc<PrometheusMetrics>,
+        status: StatusCode,
+        path: String,
+        method: Method,
+    }
 
-#[doc(hidden)]
-#[pin_project(PinnedDrop)]
-pub struct StreamLog<B> {
-    #[pin]
-    body: B,
-    size: usize,
-    clock: Instant,
-    inner: Arc<PrometheusMetrics>,
-    status: StatusCode,
-    path: String,
-    method: Method,
-}
 
-#[pinned_drop]
-impl<B> PinnedDrop for StreamLog<B> {
-    fn drop(self: Pin<&mut Self>) {
-        // update the metrics for this request at the very end of responding
-        self.inner
-            .update_metrics(&self.path, &self.method, self.status, self.clock);
+    impl<B> PinnedDrop for StreamLog<B> {
+        fn drop(this: Pin<&mut Self>) {
+            // update the metrics for this request at the very end of responding
+            this.inner
+                .update_metrics(&this.path, &this.method, this.status, this.clock);
+        }
     }
 }
 

